okhttp/src/main/java/io/grpc/okhttp/OkHttpClientTransport.java

```diff
@@ -509,9 +521,26 @@ public void run() {
         // network is not available during startup while another thread holding lock to send the
         // initial preface.
         try {
+          System.out.println("in serializing thread");
+          long waitStartTime = System.nanoTime();
+          barrier.await(10000000, TimeUnit.NANOSECONDS);
+          long waitEndTime = System.nanoTime();
+          System.out.println("after barrier");
+          if (waitEndTime - waitStartTime >= 10000000) {
+            startGoAway(0, ErrorCode.INTERNAL_ERROR, Status.UNAVAILABLE
+                .withDescription("Timed out waiting for second handshake thread. "
+                + "The transport executor pool may have run out of threads"));
+            if (connectedFuture != null) {
+              connectedFuture.set(null);
+            }
+            System.out.println("started goaway");
+            return;
+          }
           latch.await();
```

Comment:
I would like to understand the reason for this previous `latch`. 
as it seems, this code waits for `sendConnectionPrefaceAndSettings` execution, which happens below.
Can that be moved above this `serializingExecutor.execute(new Runnable()` ?
Have I missed anything?

Reply:
The introduction of DelayedClientTransport avoided using transports before they were READY. TransportSet is known as InternalSubchannel today. The activeTransportFuture.set() right after start() was the main case that sent RPCs to transports before they were ready.
https://github.com/grpc/grpc-java/commit/cf787bddf2fa110b2673cd671468eb7643c8ee03#diff-c993808318f59c8a206c5b4f1af4fd2b3f81a0fca3662be8966c392e3829e430R200

I feel like it took some years to be confident we had gotten rid of all the places that assumed a transport could support RPCs immediately after being started. But that was the most important change in that direction. The change was prompted by repeated memory leaks when dealing with ListenableFutures, as you can't remove listeners from futures. And in the process we fixed assigning RPCs too eagerly to transports.

(There had been some debate at the time whether we were okay with the added latency. Waiting for transport ready actually delays RPCs from being sent on a new connection by a RTT, because it waits to receive HTTP/2 SETTINGS from the server.)